### PR TITLE
Fix #5344 Add the ability to provide and use a CA file for mysql servers with `require_secure_transport=on`

### LINKED ISSUE #5344: Cannot connect to azure mysql flexible server while --require_secure_transport=ON
### 📑 I have found these related issues/pull requests

https://github.com/louislam/uptime-kuma/blob/8a432ac93776f01a5464f70e7d34e223193aad2e/server/setup-database.js#L212

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/uptime-kuma/security/policy)

### Description

Cannot connect to azure mysql flexible server: Connections using insecure transport are prohibited while --require_secure_transport=ON.

### 👟 Reproduction steps

**How to reproduce:**
* Create a mysql flexible server on azure.
* Try to connect it with required connection parameters.

### 👀 Expected behavior

**Expected result :**
* If credentials and connectivity is ok, test connection succeeds.

### 😓 Actual Behavior

**Actual result:**
* Connection fails with following message in web console.
* 
`Cannot connect to the database: Connections using insecure transport are prohibited while --require_secure_transport=ON.`
* As a debugging step if we configure `require_secure_transport=OFF` for the server the application will pass the connection test.

### 🐻 Uptime-Kuma Version

2.X-Beta

### 💻 Operating System and Arch

Docker louislam/uptime-kuma:2.0.0-beta.0

### 🌐 Browser

Brave 1.71.118

### 🖥️ Deployment Environment

- Runtime: Docker Azure webapp
- Database: Azure mysql flexible
- Filesystem used to store the database on:  NA
- number of monitors:  NA

### 📝 Relevant log output

```shell
Cannot connect to the database: Connections using insecure transport are prohibited while --require_secure_transport=ON.
```

### PR DESCRIPTION
**Pull request by GABRIELNGBTUC**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

Fixes #5344

This PR adds the following:
- A file input on the database setup UI for mariadb/mysql to optionally upload a PEM file
- The setup database endpoint has been modified to attempt to connect to the database with this PEM file when provided
- The `writeDBConfig` function has been modified to move a provided CA file (from either the environment variable or the setup through the UI) to a file name `mariadb-ca.pem` inside the data directory
- `db-config.json` can now contains a new property `caFilePath` with the path of the CA file to use when using secure connections
- The `connect` function has been modified to use secure connections when the property `caFilePath` is present in the  `db-config.json`
## Type of change

Please delete any options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] My code follows the style guidelines of this project
- [x]  I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)
image

### CODE REVIEW

File: server/database.js [hunk 1]
```diff
@@ -184,10 +184,18 @@
...
      * @typedef {string|undefined} envString
-     * @param {{type: "sqlite"} | {type:envString, hostname:envString, port:envString, database:envString, username:envString, password:envString}} dbConfig the database configuration that should be written
+     * @param {{type: "sqlite"} | {type:envString, hostname:envString, port:envString, database:envString, username:envString, password:envString, caFilePath:envString}} dbConfig the database configuration that should be written
      * @returns {void}
      */
     static writeDBConfig(dbConfig) {
+        // Move CA file to the data directory
+        if (dbConfig.caFilePath) {
+            const dataCaFilePath = path.join(Database.dataDir, "mariadb-ca.pem");
+            fs.renameSync(dbConfig.caFilePath, dataCaFilePath);
```
[the remark this answers is not in the document]
Comment (vsmanu) [reply server/database.js:194]: An alternate approach shall be , using a generic CA bundle : Eg: Mozilla. Then provide an env variable to override it, either as a path or the ca bundle content itself.

[the remark this answers is not in the document]
Comment (GABRIELNGBTUC) [reply server/database.js:194]: My first attempt at implementing this was to manage this entirely through environment variables but then I read from the contibuting.md the following:

> I personally do not like something that requires a lot of configuration before you can finally start the app. The goal is to make the Uptime Kuma installation as easy as installing a mobile app.

And

> Settings should be configurable in the frontend. Environment variables are discouraged, unless it is related to startup such as DATA_DIR

I believe this is something that should be configured through environment variables but if we want to respect the whishes of the maintainer, the user upload has to be there with all the complications it implies.

[the remark this answers is not in the document]
Comment (vsmanu) [reply server/database.js:194]: Again just opinion :) 

If we look at various feature requests for external Db, I believe it was not implemented early to keep the simplicity you quoted. A user base who may use a dedicated DB would be certainly having the knowledge and willingness to get it done, by reading docs and Configure As Code , even if there are efforts involved. 

In that context, if we add a UI element, theoretically it is cluttering AND/OR complicating the experience of the user base who needs just out of the box experience. 

On the other hand, if we do it via configuration those who need it can implement in a CAC fashion, hopefully it is efficient for those user base. 

In a nutshell , if we do it via config, we may able to stretch the functionality without affecting the core slogan of simplicity.

[the remark this answers is not in the document]
Comment (GABRIELNGBTUC) [reply server/database.js:194]: @louislam Would it be possible to have your input on this as I do agree with @vsmanu proposal.

Changes are minimal and simpler with an environment variable only configuration but this would only be possible if `Quote: Settings should be configurable in the frontend.` is not an hard requirement.

**Comment by GABRIELNGBTUC** [comment]
After doing some more changes locally to normalize the path after every single use and adding additional checks to make 100% sure the path cannot be hijacked, I am still failing to pass the static analysis there.

If anyone feel like making the change themselves, feel free to use this PR as a reference but I'm not feeling like fighting CodeQL anymore.

**Comment by GABRIELNGBTUC** [comment]
Reopening since I realized I was thinking too hard about how to fix the codeQL issue

**Comment by vsmanu** [comment]
Hi @GABRIELNGBTUC ,
Sorry, I was bit too busy during last few weeks. Seems the branch got stale now, could you please pull in the latest from master.

File: server/database.js [hunk 2]
```diff
@@ -183,11 +184,29 @@ class Database {
...
+     * @param {{type: "sqlite"} | {type:envString, hostname:envString, port:envString, database:envString, username:envString, password:envString, caFilePath:envString}} dbConfig the database configuration that should be written
      * @returns {void}
      */
     static writeDBConfig(dbConfig) {
+        // Move CA file to the data directory
+        const temporaryDirectoryPath = tmpdir();
+        if (dbConfig.caFilePath && dbConfig.caFile && dbConfig.caFilePath.startsWith(temporaryDirectoryPath)) {
+            const dataCaFilePath = path.resolve(Database.dataDir, "mariadb-ca.pem");
+            const sourcePath = fs.realpathSync(path.resolve(temporaryDirectoryPath, dbConfig.caFilePath));
+            if (dataCaFilePath.startsWith(path.resolve(Database.dataDir)) && path.resolve(sourcePath).startsWith(temporaryDirectoryPath)) {
```
Comment (vsmanu) [inline server/database.js:198]: Good to have the protection for path traversal prevention, but may be a symlink based traversal can happen. How about using `realpathSync()`  ?

Comment (vsmanu) [reply server/database.js:198]: Also for single users , this attack surface may ignored, as anyway an Admin user will be using it, but the threat is valid on shared infra setups.

File: server/setup-database.js [hunk 3]
```diff
@@ -206,13 +208,43 @@ class SetupDatabase {
...
+                    // Prevent someone from injecting a CA file path not generated by the code below
+                    if (dbConfig.caFilePath) {
+                        dbConfig.caFilePath = undefined;
+                    }
+
+                    if (dbConfig.caFile) {
+                        const base64Data = dbConfig.caFile.replace(/^data:application\/octet-stream;base64,/, "");
+                        const binaryData = Buffer.from(base64Data, "base64").toString("binary");
+                        const tempCaDirectory = fs.mkdtempSync(path.join(tmpdir(), "kuma-ca-"));
+                        dbConfig.caFilePath = path.resolve(tempCaDirectory, "ca.pem");
```
Comment (vsmanu) [inline server/setup-database.js:220]: Doesn't it make sense to check file content at the serverside as well ? 

`const fs = require('fs');

function isPemCertificate(filePath) {
    const content = fs.readFileSync(filePath, 'utf8');
    // Check for the PEM certificate header/footer
    const certPattern = /-----BEGIN CERTIFICATE-----[\s\S]+-----END CERTIFICATE-----/;
    return certPattern.test(content);
}

// Example usage
const filePath = './ca.pem';
if (isPemCertificate(filePath)) {
    console.log('The file is a PEM-encoded certificate.');
} else {
    console.log('The file is NOT a valid PEM certificate.');
}`

OR

`const x509 = require('@peculiar/x509'); // npm install @peculiar/x509
const fs = require('fs');

const pem = fs.readFileSync('ca.pem', 'utf8');
const cert = new x509.X509Certificate(pem);

console.log('Is CA:', cert.basicConstraintsCA); // true if it's a CA cert`

I prefer not to add extra deps as far as possible.

File: server/database.js [hunk 4]
```diff
@@ -183,11 +184,29 @@ class Database {
...
+                fs.renameSync(sourcePath, dataCaFilePath);
+            } else {
+                throw new Error("The file cannot contains any characters that would allow it to escape the data directory");
```
Comment (vsmanu) [inline server/database.js:201]: `throw new Error("The file cannot contain any characters that would allow it to escape the data directory.");`
